std/algorithm/iteration.d

```diff
@@ -4289,3 +4304,592 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Lazily computes the Cartesian power of $(D r) with itself
+for a number of repetitions $(D repeat).
+If the input is sorted, the product is in lexicographic order.
+For example $(D"AB".cartesianPower(2).array) returns $(D["AA", "AB", "BA", "BB"])
+
+Params:
+    r = $(REF_ALTTEXT Random access range, isRandomAccessRange, std, range, primitives)
+    repeat = number of repetitions
+
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat = 1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(repeat >= 1, "Invalid number of repetitions");
+}
+body
+{
+    import std.range: Indexed;
+    static struct CartesianPower
+    {
+    private:
+        Range _r;
+        size_t[] _state, _stateBack;
+
+        size_t _max_states;
+        size_t _pos, _posBack;
+
+    public:
+        this(Range r, size_t repeat)
+        {
+            import std.math: pow;
+
+            _r = r;
+
+            // set initial state and calculate max possibilities
+            if (r.length > 0)
+            {
+                _max_states = pow(r.length, repeat);
+                _state = new size_t[](repeat);
+                _stateBack = new size_t[](repeat);
+                _stateBack[] = r.length -1;
+            }
+        }
+
+        @property typeof(this) save()
+        {
+            typeof(this) c = this;
+            c._state = _state.dup;
+            return c;
+        }
+
+        @property Indexed!(Range, size_t[]) front()
+        {
+            import std.range : indexed;
+            return _r.indexed(_state);
+        }
+
+        void popFront()
+        {
+            // check whether we have reached the end
+            if (empty) return;
```

Comment:
Generally we use `assert(!empty);` in `popFront`, making it a logic error to call `popFront` on an empty range